### ISSUE TITLE
How to design the RealmObject  to explain dynamic jsonString

### ISSUE THREAD

**Issue by chenzj-king** [realm/realm-java#2838]
As title.  
I have some problem about dynamic jsonString to RealmOjbect.  

sample jsonStr:  

    {
	    "ffid": "573adf9ee8c9ea2743f87bae",
	    "ver": 10001,
	    "ffName": "test1",
	    "isplan": true,
	    "datas": {
	        "doctoradvice": "Tone up",
	        "referral": true,
	        "weight": 70,
	        "height": 175,
	        "createTime": 1463634121462,
	        "followupdoctor": "Doctor1",
	        "family": {
	            "father": "father1",
	            "mother": "mother1"
	        },
	        "symptom": [
	            {
	                "name": "swirl",
	                "key": "2"
	            }
	        ]
	    }
	}

This jsonString has int & long & double & Boolean & String type. In the jsonStr. ffid/ver/ffName/isPlan is the fixed field. But the datas is a dynamic jsonObject.sometime it will some field like "age":20. I think I can use ExtendBean Class like:  

    public class ExtendBean extends RealmObject {

	    private String key;
	    private String val;
		
		//getter and setter
	}

So datas Object i will convert to RealmList<ExtendBean>.  
But I can't to how explain the datas jsonStr by Gson to RealmList<ExtendBean>. And also how can I convert RealmList<ExtendBean> to datas Object.

can you guide me to solve this problem.

#### Version of Realm and tooling
Realm version(s): 0.90.1

Android Studio version: 2.1.1

Which Android version and device: 5.1

**Comment by stk1m1** [comment]
Hello @chenzj-king,

In case you have two different keys and values in JSON being filled up separately, I'd suggest you have two fields in your model such that one is `age` and another is `datas`. When one is missing (e.g. you have `ages` instead of `datas`), you'll have `ages` field set with value `20` and vice versa.

If you have one key `datas` and its content is changing, (e.g. sometime `20`, other times JSON data), then you'd have to parse the JSON data and map the data with object yourself. 

We have [documented](https://realm.io/docs/java/latest/#json) and provided [assorted examples](https://github.com/realm/realm-java/tree/master/examples/jsonExample) for handling JSON data.

**Comment by chenzj-king** [comment]
@stk1m1  Thank you for you answer.  
I find a way didn't use gson can explain the jsonStr to RealmList<Extend>. 

    public class Extend extends RealmObject {
	    @Required
	    String key;
	    @Required
	    String value;
	    @Required
	    String type;
	}


I use the JsonObject to explain the JsonStr in datas.  

    "datas": {
        "doctoradvice": "Tone up",
        "referral": true,
        "weight": 70,
        "height": 175,
        "createTime": 1463634121462,
        "followupdoctor": "Doctor1",
        "family": {
            "father": "father1",
            "mother": "mother1"
        },
        "symptom": [
            {
                "name": "swirl",
                "key": "2"
            }
        ]
    }

Our project only dynamic in datas Object. And I only should save the first floor message. The other floor I will save as String. like JsonObject/JsonArray. I will xx.toString and as extend.val.  

    String jsonStr = "{\n" +
            "    \"int\": 0,\n" +
            "    \"long\": 1000000000000000000,\n" +
            "    \"double\": 10000.2,\n" +
            "    \"boolean\": false,\n" +
            "    \"string\": \"testString\",\n" +
            "    \"obj\": {\n" +
            "        \"testObj\": \"obj\"\n" +
            "    },\n" +
            "    \"array\": [\n" +
            "        \"array0\",\n" +
            "        \"array1\"\n" +
            "    ]\n" +
            "}";

    JSONObject jsonObject = null;
    try {
        jsonObject = new JSONObject(jsonStr);

        RealmList<Extend> realmList = new RealmList<>();

        Iterator<String> keys = jsonObject.keys();

        while (keys.hasNext()) {

            Extend extend = new Extend();

            String key = keys.next();
            Object obj = jsonObject.get(key);

            extend.key = key;
            extend.value = String.valueOf(obj);

            if (obj instanceof Integer) {
                extend.type = "Integer";
            } else if (obj instanceof Long) {
                extend.type = "Long";
            } else if (obj instanceof Double) {
                extend.type = "Double";
            } else if (obj instanceof Boolean) {
                extend.type = "Boolean";
            } else if (obj instanceof String) {
                extend.type = "String";
            } else if (obj instanceof JSONObject) {
                extend.type = "JSONObject";
            } else if (obj instanceof JSONArray) {
                extend.type = "JSONArray";
            }

            realmList.add(extend);
        }
    } catch (JSONException e) {
        e.printStackTrace();
    }

When I need to serialize to jsonStr. I will get the extend.type and jsonObject.put(extend.key,Integer/Double/Boolean/String.parseXxType(extend.val))  

I think it can work well but it's not cool.

**Comment by stk1m1** [comment]
@chenzj-king 

I can see your hassle with the JSON model. You don't have to go through all the manual mapping like above if the `datas` part 
```json
"datas": {
    "doctoradvice": "Tone up",
    "referral": true,
    "weight": 70,
    "height": 175,
    "createTime": 1463634121462,
    "followupdoctor": "Doctor1",
    "family": {
        "father": "father1",
        "mother": "mother1"
    },
    "symptom": [
        {
            "name": "swirl",
            "key": "2"
        }
    ]
}
```
is consistently represented in the same JSON model. (i.e. the content of `datas` does not randomly change to `ages:20`) 

The `datas` portion could be defined as a clean, separate child `RealmObject` and, in fact, it is a good API design practice to keep one key to represent one data model.

**Comment by chenzj-king** [comment]
@stk1m1 

Actually. I didn't now the datas have what field. The field datas is get by h5. Our app in hybrid.  I just save the data. And post to server or post to h5. But in the future. I need to search some itemData. So i need to save as RealmList<Extend> like above. and then i can use query to like  

    realmQuery.equalTo("mExtendBeanRealmList.val", education);

It can easy to implementation in now Realm Api.  
 
In conclusion. Can you give me some advice about how to save the datas JsonObject. I think the Extend class can run well. although it's not cool.

**Comment by stk1m1** [comment]
@chenzj-king 

You can define another `RealmObject` for JSONObject. Then persist it and link it to a child object to the parent `ExtendBean` model when you are sure `datas` part could be parsed as JSON.

**Comment by chenzj-king** [comment]
@stk1m1 
Yes. I has do that. It work well. Just need to write a util class to jsonStr to RealmList<ExtendBean> or RealmList<ExtendBean> to jsonStr.  And it also can RealmQuery<E> to query first layer data.
